### ISSUE TITLE
Naming/MemoizedInstanceVariableName Cop doesn't work with methods that have a leading underscore

### ISSUE THREAD

**Issue by ashmaroli** [rubocop-hq/rubocop#6084]
[`jekyll/convertible.rb`](https://github.com/jekyll/jekyll/blob/70f4d4908de10239b4774210a0e720b7df3b32f9/lib/jekyll/convertible.rb#L241-L245) has a private method named `_renderer` which memoizes using the ivar `@_renderer` as follows:
```ruby
    private

    def _renderer
      @_renderer ||= Jekyll::Renderer.new(site, self)
    end
```
--------

## Expected behavior

Rubocop should not flag the above code when I configure the `Naming/MemoizedInstanceVariableName` with the provided directives for `EnforcedStyleForLeadingUnderscores:`

## Actual behavior

1. Configured with **`EnforcedStyleForLeadingUnderscores: disallowed`**
```
Naming/MemoizedInstanceVariableName:
  EnforcedStyleForLeadingUnderscores: disallowed
```
```
Offenses:

lib/jekyll/convertible.rb:244:7: C: Naming/MemoizedInstanceVariableName: Memoized variable
@_renderer does not match method name _renderer. Use @_renderer instead.
      @_renderer ||= Jekyll::Renderer.new(site, self)
      ^^^^^^^^^^
```

2. Configured with **`EnforcedStyleForLeadingUnderscores: required`**
```
Naming/MemoizedInstanceVariableName:
  EnforcedStyleForLeadingUnderscores: required
```
```
Offenses:

lib/jekyll/convertible.rb:244:7: C: Naming/MemoizedInstanceVariableName: Memoized variable
@_renderer does not match method name _renderer. Use @_renderer instead.
      @_renderer ||= Jekyll::Renderer.new(site, self)
      ^^^^^^^^^^
```

3. Configured with **`EnforcedStyleForLeadingUnderscores: optional`**
```
Naming/MemoizedInstanceVariableName:
  EnforcedStyleForLeadingUnderscores: optional
```
```
Offenses:

lib/jekyll/convertible.rb:244:7: C: Naming/MemoizedInstanceVariableName: Memoized variable
@_renderer does not match method name _renderer. Use @_renderer instead.
      @_renderer ||= Jekyll::Renderer.new(site, self)
      ^^^^^^^^^^
```
All of the above scenario flags the variable `@_renderer` and **suggests** using the same variable `@_renderer` instead.

---

## RuboCop version

```
$ rubocop -V
0.58.0
```

**Comment by Drenmi** [comment]
Problem is the cop does not expect methods starting with `_`.

**Comment by kenman345** [comment]
@Drenmi I am trying to understand so maybe I can help fix this issue. Why would a setting for the cop exist that is specifically for handling underscores yet it is not expecting to see underscores? How does code like that get added? 

Can you help direct me towards where the issue might be that I can try and resolve it? I am not familiar with rubocop yet and it would save me a little time to get some direction on this.

**Comment by kenman345** [comment]
Nevermind, I think I found the area, but still confused why a new option wouldnt actually work or have a test proving it works before its added to a release.

**Comment by Drenmi** [comment]
@kenman345 The setting expects a prefix underscore in the *variable name*, but doesn't expect it in the method name. I.e. it expects this:

```ruby
def foo
  @_foo ||= ...
end
```

and it checks it by stripping the underscore from the variable name and comparing it to the unaltered method name. So in the case of an underscore method, it compares `foo` (variable name) to `_foo` (method name.)

**Comment by ashmaroli** [comment]
@bbatsov This issue has not been resolved as expected.
It is reproducible with the latest `rubocop:master`

**Comment by timoschilling** [comment]
This is still an issue, `0.58.2` didn't fix this issue in my case:
```ruby
    def _warden_user
      @_warden_user ||= _warden.authenticate(scope: :user)
    end
```
```
authentication.rb:40:7: C: Naming/MemoizedInstanceVariableName: Memoized variable @_warden_user does not match method name _warden_user. Use @_warden_user instead.
      @_warden_user ||= _warden.authenticate(scope: :user)
      ^^^^^^^^^^^^^
```

**Comment by kenman345** [comment]
That’s a bit different as your method name is in fact different from the
variable name. This is not a cop purely to enforce underscores for
memoization

**Comment by kenman345** [comment]
Sorry, I think that it’s using the memoization statement as the method def
and variable names there. Odd right? Try the optional enforcement and see
what happens. Can you provide your repo link and I can give it a try?
Otherwise I’ll just add in a method like that and see what’s being matched.

**Comment by timoschilling** [comment]
simply use this:
```ruby
class Test
  def ok
    @ok ||= 1
  end

  def _not_ok
    @_not_ok ||= 1
  end
end
```

```bash
$ bundle exec rubocop -v
0.57.2
$ bundle exec rubocop
Inspecting 1 files
.

1 files inspected, no offenses detected
```
```bash
$ bundle exec rubocop -v
0.58.2
$ bundle exec rubocop
Inspecting 1 files
C

Offenses:

test.rb:7:5: C: Naming/MemoizedInstanceVariableName: Memoized variable @_not_ok does not match method name _not_ok. Use @_not_ok instead.
    @_not_ok ||= 1
    ^^^^^^^^

1 files inspected, 1 offense detected
```

**Comment by kenman345** [comment]
I’m not at my Pc right now but that seems off. That’s basically the exact
thing we have in the spec tests for my PR.

Is this with Ash’s recommendation to configure it? Just want to know
everything before testing when I have the chance

**Comment by timoschilling** [comment]
It's without any config.

**Comment by timoschilling** [comment]
Now I have read https://github.com/rubocop-hq/rubocop/blob/4ed9aab2b2ba8f1a004c58e5765940540742623d/manual/cops_naming.md#namingmemoizedinstancevariablename.
And I still think it is a bug.

EnforcedStyleForLeadingUnderscores: disallowed (default)
Means the variable must be qual the method name.
This is the case in my code and config. But Rubocop detects a offense. That is a bug.

Sure it works with `optional` but it should work with `required` too.

And even the error message shows up a bug:
> Memoized variable @_not_ok does not match method name _not_ok. Use @_not_ok instead.
The given and the suggested var name are the same.

**Comment by timoschilling** [comment]
I have checked the tests and my case isn't covered, I will provide a PR with the example for that case.
